prow/spyglass/viewers/buildlog/buildlog_viewer.go

```diff
@@ -0,0 +1,206 @@
+		Title:    title,
+		Priority: priority,
+	}, ViewHandler)
+}
+
+// LogViewerRequest contains the operations to perform on each log artifact
+type LogViewerRequest struct {
+	// Requests is a map of log artifact names to operations to perform on that log artifact
+	Requests map[string]LogViewData `json:"requests"`
+}
+
+// LogViewData holds the current state of a log artifact's view and the operation to be performed on it.
+// Valid operations: more, all, ""
+type LogViewData struct {
+	CurrentChunks int    `json:"currentChunks"`
+	TotalChunks   int    `json:"totalChunks"`
+	Operation     string `json:"operation"`
+}
+
+//LogArtifactView holds a single log file's view
+type LogArtifactView struct {
+	ArtifactName          string
+	ArtifactLink          string
+	ViewName              string
+	ViewMethodDescription string
+	LogLines              []string
+}
+
+// BuildLogsView holds each log file view
+type BuildLogsView struct {
+	LogViews           []LogArtifactView
+	GetAllRequests     map[string]LogViewerRequest
+	GetMoreRequests    map[string]LogViewerRequest
+	RawGetAllRequests  map[string]string
+	RawGetMoreRequests map[string]string
+}
+
+// ViewHandler creates a view for a build log (or multiple build logs)
+func ViewHandler(artifacts []viewers.Artifact, raw string) string {
+	buildLogsView := BuildLogsView{
+		LogViews:           []LogArtifactView{},
+		GetAllRequests:     make(map[string]LogViewerRequest),
+		GetMoreRequests:    make(map[string]LogViewerRequest),
+		RawGetAllRequests:  make(map[string]string),
+		RawGetMoreRequests: make(map[string]string),
+	}
+	viewReq := LogViewerRequest{
+		Requests: make(map[string]LogViewData),
+	}
+	if err := json.Unmarshal([]byte(raw), &viewReq); err != nil {
+		logrus.WithError(err).Error("Error unmarshaling log view request data.")
+	}
+
+	getAllReq := LogViewerRequest{
+		Requests: make(map[string]LogViewData),
+	}
+
+	getMoreReq := LogViewerRequest{
+		Requests: make(map[string]LogViewData),
+	}
+
+	var desc string
+	var chunks int
+	var totalChunks int
+	// Read log artifacts and construct template structs
+	for _, a := range artifacts {
+		viewData, ok := viewReq.Requests[a.JobPath()]
+		if ok {
+			chunks = viewData.CurrentChunks
+			totalChunks = viewData.TotalChunks
+			switch viewData.Operation {
+			case "more":
+				if chunks < viewData.TotalChunks {
+					chunks++
+				}
+			case "all":
+				chunks = viewData.TotalChunks
+			case "":
+			default:
+				logrus.Error("Invalid BuildLogViewer operation provided.")
+			}
+		} else {
+			totalChunks = int(math.Ceil(float64(a.Size()) / float64(viewChunkSize)))
+			chunks = 1
+		}
+		logLines, err := viewers.LastNLines(a, int64(nLines*chunks))
+		desc = fmt.Sprintf("viewing last %d lines", nLines*chunks)
+		if err != nil {
+			logrus.WithError(err).Error("Last N lines failed.")
+			desc = fmt.Sprintf("viewing first %d bytes", viewChunkSize*chunks)
+			if err == viewers.ErrUnsupportedOp { // Try a different read operation
```

Comment:
Is there a specific advantage to not exporting the error?

Reply:
both are used in the go stdlib. not exporting the type is useful for EG the syscall / OS stuff which has some common functionality abstracted over system specific implementations. I'd think exporting would be preferred here actually.

https://golang.org/pkg/os/#pkg-variables
vs
https://golang.org/pkg/os/#IsNotExist

The latter error type uses `IsNotExist` because the `stat` errors are going to contain system specific data in them.
`ErrClosed` etc. are just exported on the other hand because operating on file handles is relatively portable as are the errors and no extra information is contained by the error.